javascript/ql/lib/semmle/javascript/frameworks/TypeORM.qll

```diff
@@ -0,0 +1,188 @@
+    result = dataMapperQueryBuilder() or
+    result = activeRecordQueryBuilder()
+  }
+
+  /**
+   * Gets The Brackets that are SQL Subqueries equivalent
+   */
+  API::Node brackets() {
+    result =
+      API::moduleImport("typeorm")
+          .getMember(["Brackets", "NotBrackets"])
+          .getParameter(0)
+          .getParameter(0)
+  }
+
+  /**
+   * Gets any Successor node of Brackets, NotBrackets
+   */
+  API::Node getASuccessorOfBrackets() {
+    result = brackets() or
+    result = getASuccessorOfBrackets().getAMember() or
+    result = getASuccessorOfBrackets().getAParameter() or
+    result = getASuccessorOfBrackets().getUnknownMember() or
+    result = getASuccessorOfBrackets().getReturn() or
+    result = getASuccessorOfBrackets().getInstance()
+  }
+
+  /**
+   * Gets any Successor node of createQueryBuilder
+   */
+  API::Node getASuccessorOfBuilderInstance() {
+    result = queryBuilderInstance() or
+    result = getASuccessorOfBuilderInstance().getAMember() or
+    result = getASuccessorOfBuilderInstance().getAParameter() or
+    result = getASuccessorOfBuilderInstance().getUnknownMember() or
+    result = getASuccessorOfBuilderInstance().getReturn() or
+    result = getASuccessorOfBuilderInstance().getInstance()
+  }
+
+  /**
+   *  Gets functions responsible for select expressions
+   *  `orderBy` is not injectable in TypeORM, if we want to write a filter we should specify a DataSource parameter string value,
+   *  which mostly is taken from config files and we will loose many sinks,
+   *  Also many application support multiple DBMSs besides TypeORM,
+   *  Also Consider it that `Order By` clause is one of the most popular injectable sinks
+   */
+  string selectExpression() {
+    result =
+      [
+        "select", "addSelect", "where", "andWhere", "orWhere", "having", "orHaving", "andHaving",
+        "orderBy", "addOrderBy", "distinctOn", "groupBy", "addCommonTableExpression",
+        "leftJoinAndSelect", "innerJoinAndSelect", "leftJoin", "innerJoin", "leftJoinAndMapOne",
+        "innerJoinAndMapOne", "leftJoinAndMapMany", "innerJoinAndMapMany", "orUpdate", "orIgnore",
+        "values", "set"
+      ]
+  }
+
+  /**
+   * Gets functions that return results
+   */
+  string queryBuilderResult() {
+    result = ["getOne", "getOneOrFail", "getMany", "getRawOne", "getRawMany", "stream"]
+  }
+
+  /**
+   *  A call to some successor functions of TypeORM `createQueryBuilder` function which are dangerous
+   */
+  private class QueryBuilderCall extends DatabaseAccess, API::CallNode {
+    API::Node typeOrmNode;
+
+    QueryBuilderCall() {
+      typeOrmNode = getASuccessorOfBuilderInstance() and
+      this = typeOrmNode.asSource()
+      or
+      // I'm doing following because `this = TypeORMNode.asSource()`
+      // don't let me to get a member in getAQueryArgument
+      typeOrmNode = getASuccessorOfBrackets() and
+      typeOrmNode.getMember(selectExpression()).getACall() = this
```

Comment:
I suspect this will change when you change `getASuccessorOfBuilderInstance()`, because this looks strange.